src/printf/printf.c

```diff
@@ -701,55 +773,63 @@ static struct scaling_factor update_normalization(struct scaling_factor sf, floa
 
 static struct floating_point_components get_normalized_components(bool negative, printf_size_t precision, floating_point_t non_normalized, struct scaling_factor normalization, int floored_exp10)
 {
-  struct floating_point_components components;
-  components.is_negative = negative;
-  floating_point_t scaled = apply_scaling(non_normalized, normalization);
+  floating_point_t scaled;
+  bool close_to_representation_extremum;
+  scaled = apply_scaling(non_normalized, normalization);
 
-  bool close_to_representation_extremum = ( (-floored_exp10 + (int) precision) >= FP_TYPE_MAX_10_EXP - 1 );
+  close_to_representation_extremum = ( (-floored_exp10 + (int) precision) >= FP_TYPE_MAX_10_EXP - 1 );
   if (close_to_representation_extremum) {
-    // We can't have a normalization factor which also accounts for the precision, i.e. moves
-    // some decimal digits into the mantissa, since it's unrepresentable, or nearly unrepresentable.
-    // So, we'll give up early on getting extra precision...
+    /*
+     * We can't have a normalization factor which also accounts for the precision, i.e. moves
+     * some decimal digits into the mantissa, since it's unrepresentable, or nearly unrepresentable.
+     * So, we'll give up early on getting extra precision...
+     */
     return get_components(SIGN(negative, scaled), precision);
+  } else {
+    int_fast64_t integral = (int_fast64_t) scaled;
+    floating_point_t remainder = non_normalized - unapply_scaling((floating_point_t) integral, normalization);
+    floating_point_t prec_power_of_10 = powers_of_10[precision];
+    struct scaling_factor account_for_precision = update_normalization(normalization, prec_power_of_10);
+    floating_point_t scaled_remainder = apply_scaling(remainder, account_for_precision);
+    floating_point_t rounding_threshold = 0.5;
+    struct floating_point_components components;
+
+    components.integral = integral;
+    components.is_negative = negative;
+    components.fractional = (int_fast64_t) scaled_remainder; /* when precision == 0, the assigned value should be 0 */
+    scaled_remainder -= (floating_point_t) components.fractional; /* when precision == 0, this will not change scaled_remainder */
+
+    components.fractional += (scaled_remainder >= rounding_threshold);
+    if (scaled_remainder == rounding_threshold) {
+      /* banker's rounding: Round towards the even number (making the mean error 0) */
+      components.fractional &= ~((int_fast64_t) 0x1);
+    }
+    /*
+    * handle rollover, e.g. the case of 0.99 with precision 1 becoming (0,100),
+    * and must then be corrected into (1, 0).
+    * Note: for precision = 0, this will "translate" the rounding effect from
+    * the fractional part to the integral part where it should actually be
+    * felt (as prec_power_of_10 is 1)
+    */
```

Comment:
```suggestion
    /*
     * handle rollover, e.g. the case of 0.99 with precision 1 becoming (0,100),
     * and must then be corrected into (1, 0).
     * Note: for precision = 0, this will "translate" the rounding effect from
     * the fractional part to the integral part where it should actually be
     * felt (as prec_power_of_10 is 1)
     */
```